dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelCrit.java

```diff
@@ -23,59 +30,95 @@ public RelCrit(){
-                        }
+        if(maxTupleSet != null) {
+            return maxTupleSet;
+        }
+        maxTupleSet = new TupleSet();
+        minTupleSet = new TupleSet();
+        ExecutionAnalysis exec = analysisContext.get(ExecutionAnalysis.class);
+        for (Thread thread : task.getProgram().getThreads()) {
+            // assume order by cId
+            // assume cId describes a topological sorting over the control flow
+            List<Event> locks = reverse(thread.getCache().getEvents(FilterBasic.get(RCU_LOCK)));
+            for (Event unlock : thread.getCache().getEvents(FilterBasic.get(RCU_UNLOCK))) {
+                // iteration order assures that all intermediaries were already iterated
+                for (Event lock : locks) {
+                    if (unlock.getCId() < lock.getCId() ||
+                            exec.areMutuallyExclusive(lock, unlock) ||
+                            concat(minTupleSet.getByFirst(lock).stream().map(Tuple::getSecond),
+                                            minTupleSet.getBySecond(unlock).stream().map(Tuple::getFirst))
+                                    .anyMatch(e -> exec.isImplied(lock, e) || exec.isImplied(unlock, e))) {
+                        continue;
+                    }
+                    boolean noIntermediary = maxTupleSet.getBySecond(unlock).stream()
+                                    .allMatch(t -> exec.areMutuallyExclusive(lock, t.getFirst())) &&
+                            maxTupleSet.getByFirst(lock).stream()
+                                    .allMatch(t -> exec.areMutuallyExclusive(t.getSecond(), unlock));
+                    Tuple tuple = new Tuple(lock, unlock);
+                    maxTupleSet.add(tuple);
+                    if (noIntermediary) {
+                        minTupleSet.add(tuple);
                     }
                 }
             }
-            removeMutuallyExclusiveTuples(maxTupleSet);
         }
         return maxTupleSet;
     }
 
-    // TODO: Not the most efficient implementation
-    // Let's see if we need to keep a reference to a thread in events for anything else, and then optimize this method
+    @Override
+    public void addEncodeTupleSet(TupleSet tuples) {
+        LinkedList<Tuple> queue = new LinkedList<>(intersection(tuples, maxTupleSet));
+        while (!queue.isEmpty()) {
+            Tuple tuple = queue.remove();
+            if (!encodeTupleSet.add(tuple)) {
+                continue;
+            }
+            Event lock = tuple.getFirst();
+            Event unlock = tuple.getSecond();
+            for (Tuple t : maxTupleSet.getBySecond(unlock)) {
+                if (isOrdered(lock, t.getFirst(), unlock)) {
+                    queue.add(t);
+                }
+            }
+            for (Tuple t : maxTupleSet.getByFirst(lock)) {
+                if (isOrdered(lock, t.getSecond(), unlock)) {
+                    queue.add(t);
+                }
+            }
+        }
+    }
+
```

Comment:
Just a little remark on this.
The `encodeSet` has a single purpose: Make sure that *at least* those relationships are encoded because someone needs them.
Now this makes all of the following approaches valid:
- Whenever someone enlarges the `encodeSet`, add whatever you additionally need to the `encodeSet` (as you do right now).
- On encoding time, enlarge the `encodeSet` as you need (i.e. do it once as a preprocessing/postprocessing step). It is sound to enlarge the `encodeSet` arbitrarily.
- On encoding time, just encode more than is forced by the `encodeSet`. It is sound to encode more than necessary (it would be nice to reflect this in the `encodeSet` though, so I would prefer the second point over this one).

I don't know which of these solutions fits best here, but I just want to mention them.

Reply:
For a relation without dependencies, it should not matter when this closure is computed.  I rather stick to the first approach, as to keep modifications to `encodeTupleSet` in one phase.